sources/representation/qt/mainform.cpp

```diff
@@ -178,43 +184,95 @@ void MainForm::startGameLoop(int id, QString map)
     qint64 max_process_time = 0;
-
-    while (true)
+    
+    if (!cap)
     {
-        ++fps_counter;
-
-        process_performance.start();
-        GetRepresentation().Process();
-        GetScreen().Swap();
-        qint64 process_time = process_performance.nsecsElapsed();
-        if (process_time > max_process_time)
+        while (true)
         {
-            max_process_time = process_time;
-        }
+            ++fps_counter;
 
-        if (isHidden())
-        {
-            break;
-        }
+            process_performance.start();
+            GetRepresentation().Process();
+            GetScreen().Swap();
+            qint64 process_time = process_performance.nsecsElapsed();
+            if (process_time > max_process_time)
+            {
+                max_process_time = process_time;
+            }
 
-        if (fps_timer.elapsed() > 1000)
+            if (isHidden())
+            {
+                break;
+            }
+
+            if (fps_timer.elapsed() > 1000)
+            {
+                addSystemText("FPS", "FPS: " + QString::number(fps_counter));
+                addSystemText(
+                    "{Perf}Represent",
+                    "Represent max: "
+                    + QString::number(max_process_time / 1e6)
+                    + " ms");
+                qint64 mutex_ns = GetRepresentation().GetPerformance().mutex_ns;
+                addSystemText(
+                    "{Perf}RepresentMutex",
+                    "Represent mutex lock max: "
+                    + QString::number(mutex_ns / 1e6)
+                    + " ms");
+                GetRepresentation().ResetPerformance();
+                max_process_time = 0;
+                fps_timer.restart();
+                fps_counter = 0;
+            }
+        }
+    }
+    else
+    {
+        int delta_time = 0;
+        float fps_limit = 1000 / cap;
+        QTime delta;
+        delta.start();
+        while (true)
         {
-            addSystemText("FPS", "FPS: " + QString::number(fps_counter));
-            addSystemText(
-                "{Perf}Represent",
-                  "Represent max: "
-                + QString::number(max_process_time / 1e6)
-                + " ms");
-            qint64 mutex_ns = GetRepresentation().GetPerformance().mutex_ns;
-            addSystemText(
-                "{Perf}RepresentMutex",
-                  "Represent mutex lock max: "
-                + QString::number(mutex_ns / 1e6)
-                + " ms");
-            GetRepresentation().ResetPerformance();
-            max_process_time = 0;
-            fps_timer.restart();
-            fps_counter = 0;
+            process_performance.start();
+            delta_time += delta.restart();
+            if (delta_time > fps_limit)
+            {
+                delta_time -= fps_limit;
+                ++fps_counter;
+                GetRepresentation().Process();
+                GetScreen().Swap();
+            }
```

Comment:
Now your code will just constantly perform `delta_time += delta.restart();` operation and `if` check afterwards. That does not give any time for other activities either.  

1. The core network code works in a different thread, so that will not affect on him.
2. The core game code works in a different thread and communicates directrly with the network thread.
3. All other activities are based on the Qt event loop of the representation thread. That means that they are totally fine with some delays.

Some issues can appear only when `max_fps` is really low (less than ~10 fps or smth). It may be reasonable to not sleep too long in such cases, but instead split the sleep phase:
```
while (WaitConditionIsTrue())
{
    QThread::msleep(40);
    QCoreApplication::processEvents(QEventLoop::AllEvents, 20);
}
```